src/lib/mina_base/transaction_logic.ml

```diff
@@ -1142,316 +1215,306 @@ module Make (L : Ledger_intf) : S with type ledger := L.t = struct
+              type t = L.t
+
+              let to_yojson l = L.merkle_root l |> Ledger_hash.to_yojson
+            end in
+            let module J = struct
+              type t =
+                ( Inputs.Parties.party list
+                , Token_id.t
+                , Amount.t
+                , L.t
+                , bool
+                , unit )
+                Parties_logic.Local_state.t
+              [@@deriving to_yojson]
+            end in
+            `Did_not_succeed
+        | exception e ->
+            `Error e
+        | s ->
+            step_all s
+    in
+    let init ~will_succeed : Inputs.Global_state.t * _ =
+      let parties =
+        let p = c.fee_payer in
+        { Party.authorization= Control.Signature p.authorization
+        ; data= {p.data with predicate= Party.Predicate.Nonce p.data.predicate}
+        }
+        :: c.other_parties
+      in
+      M.start
+        {parties; will_succeed; protocol_state_predicate= c.protocol_state}
+        {perform}
+        ( {protocol_state= state_view; ledger; fee_excess= Amount.zero}
+        , { parties= []
+          ; transaction_commitment= ()
+          ; token_id= Token_id.invalid
+          ; excess= Currency.Amount.zero
+          ; ledger
+          ; success= true
+          ; will_succeed } )
+    in
+    let accounts () =
+      List.map original_account_states
+        ~f:(Tuple2.map_snd ~f:(Option.map ~f:snd))
+    in
+    match step_all (init ~will_succeed:true) with
+    | `Error e ->
+        Error (Error.of_exn e)
+    | `Ok (global_state, local_state) ->
+        Ok
+          ( { accounts= accounts ()
+            ; command=
+                { With_status.data= c
+                ; status=
+                    (* TODO *)
+                    Applied
+                      ( { fee_payer_account_creation_fee_paid= None
+                        ; receiver_account_creation_fee_paid= None
+                        ; created_token= None }
+                      , { fee_payer_balance= None
+                        ; source_balance= None
+                        ; receiver_balance= None } ) } }
+          , (local_state, global_state.fee_excess) )
+    | `Did_not_succeed -> (
+        (* Restore the previous state *)
+        List.fold original_account_states ~init:[] ~f:(fun acc (id, a) ->
+            match a with
+            | None ->
+                id :: acc
+            | Some (loc, a) ->
+                L.set ledger loc a ; acc )
+        |> L.remove_accounts_exn ledger ;
+        match step_all (init ~will_succeed:false) with
+        | `Error e ->
+            Error (Error.of_exn e)
+        | `Did_not_succeed ->
+            assert false
+        | `Ok (global_state, local_state) ->
+            Ok
+              ( { accounts= accounts ()
+                ; command=
+                    { With_status.data= c
+                    ; status=
+                        Failed
+                          ( Predicate (* TODO *)
```

Comment:
I think these can be done in a follow-up PR (as with the similar failure reason stuff), but commenting anyway so that we remember to open an issue for it.

Reply:
:+1:  https://github.com/MinaProtocol/mina/issues/9060